long-ride-alerts/src/main/java/org/apache/flink/training/exercises/longrides/LongRidesExercise.java

```diff
@@ -98,17 +101,42 @@ public static void main(String[] args) throws Exception {
     @VisibleForTesting
     public static class AlertFunction extends KeyedProcessFunction<Long, TaxiRide, Long> {
 
+        private ValueState<TaxiRide> firstEvent;
+
         @Override
-        public void open(Configuration config) throws Exception {
-            throw new MissingSolutionException();
+        public void open(Configuration config) {
+            firstEvent = getRuntimeContext().getState(new ValueStateDescriptor<>("taxi ride event", TaxiRide.class));
         }
 
         @Override
-        public void processElement(TaxiRide ride, Context context, Collector<Long> out)
-                throws Exception {}
+        public void processElement(TaxiRide currentRide, Context context, Collector<Long> out)
+                throws Exception {
+            TaxiRide firstRide = firstEvent.value();
+            if (firstRide == null) {
+                firstEvent.update(currentRide);
+                if (currentRide.isStart) {
+                    context.timerService().registerEventTimeTimer(currentRide.getEventTimeMillis() + Duration.ofHours(2).toMillis());
+                }
+                return;
+            }
+            if (currentRide.isStart) {
+                if (Duration.between(currentRide.eventTime, firstRide.eventTime).compareTo(Duration.ofHours(2)) > 0) {
+                    out.collect(currentRide.rideId);
+                }
+            } else {
+                context.timerService().deleteEventTimeTimer(firstRide.getEventTimeMillis() + Duration.ofHours(2).toMillis());
+                if (Duration.between(firstRide.eventTime, currentRide.eventTime).compareTo(Duration.ofHours(2)) > 0) {
+                    out.collect(currentRide.rideId);
+                }
+            }
+            firstEvent.clear();
+        }
 
         @Override
-        public void onTimer(long timestamp, OnTimerContext context, Collector<Long> out)
-                throws Exception {}
+        public void onTimer(long timestamp, OnTimerContext context, Collector<Long> out) throws Exception {
+            TaxiRide taxiRide = firstEvent.value();
+            out.collect(taxiRide.rideId);
+            firstEvent.clear();
```

Comment:
What happens when start event arrives at t=0, then 3 hours elapse, and then the end event arrives?

Reply:
The start event will be saved in the ValueState, a timer will be registered. After 2 hours passed the timer will trigger and onTimer will execute, clearing the state. When the end event arrives, state is null, so it will be stored in ValueState.   I think this event will not be cleared, because I'm registering a timer only for start events. I think I need to register the timer regardless of the type of event (start or stop), so remove the condition on line 115. Is that correct?

Reply:
That is the problem, not sure about the solution. I have to think about it.

Reply:
I pushed a potential solution here, dealing with the scenario described by registering a timer for end event as well, and doing the clean-up when onTimer() is called. I'm also cancelling the timer for both types of events when the second event of the pair is received. With this approach, I also need to collect results on onTimer() only for start events to prevent the same rideId being collected twice.

Reply:
Sounds plausible, though I cannot confirm it without a unit test.

What I thought of is this: the environment is configured to use a watermark strategy of
```
WatermarkStrategy<TaxiRide> watermarkStrategy =
                WatermarkStrategy.<TaxiRide>forBoundedOutOfOrderness(Duration.ofSeconds(60))
                        .withTimestampAssigner(
                                (ride, streamRecordTimestamp) -> ride.getEventTimeMillis());
        // create the pipeline
        rides.assignTimestampsAndWatermarks(watermarkStrategy)
```
Watermarks are periodic timestamps included with the stream and emitted periodically if configured. I am not sure if this was covered in the first courses, or if they will be mentioned later when discussing Windows.

The expectation is: ride start events will be followed by ride end events, although they may still arrive out of order.
Note that the strategy is selected to allow a start event to arrive at most 60 seconds after the end event. Flink does not do anything explicit (like dropping the event) in this case when .process() methods are used.

So, when getting the end ride event and the firstEvent state is empty, it means either:
* the start event happend more than two hours ago, and the state was cleared, so nothing to do
* the start event will arrive up to one minute late, so we should keep the end event only until then

I think your second time that is registered for end events handles the second point, although you set it to 2 hours and not 1 minute, but I think the state will be cleared in the end either way.

Reply:
Thanks for taking the time to review this. Yes, for the second scenario you mentioned we could probably set a different timer of 1 minute, that would make more sense, given that the watermark strategy was already defined with 1 minute bound. For the first scenario, I don't think you can differentiate between an end event that arrives after more than two hours and an end event that arrived before the start event unless you keep track somehow of the start events even after the 2 hour timer triggers. I think that might be outside of the scope for this lab, but its a good exercise to think of all the possible scenarios.